### PR TITLE
Support S3 range read with buffer cache

### PR DESCRIPTION
**Pull request by LuQQiu**
### What changes are proposed in this pull request?

Please outline the changes and how this PR fixes the issue.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java [hunk 1]
```diff
@@ -79,23 +78,15 @@ public S3AInputStream(String bucketName, String key, AmazonS3 client,
...
-      openStream();
+    if (read(mSingleByteHolder) == -1) {
+      return -1;
     }
-    int value = mIn.read();
-    if (value != -1) { // valid data read
-      mPos++;
-    }
-    return value;
+    return mSingleByteHolder[0];
```
Comment (jja725) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:89]: why return the data in the holder?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 2]
```diff
@@ -1471,6 +1471,16 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.SERVER)
           .build();
+  public static final PropertyKey UNDERFS_S3_IN_STREAM_BUFFER_SIZE =
```
Comment (jja725) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:1474]: nit: better remove this property after we did micro bench and decide what's the best size for most cases. We have too many properties...

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java [hunk 3]
```diff
@@ -105,56 +96,59 @@ public int read(byte[] b) throws IOException {
 
   @Override
   public int read(byte[] b, int offset, int length) throws IOException {
-    if (length == 0) {
-      return 0;
+    S3Object object = getObject(mPos, mPos + length);
```
Comment (jja725) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:99]: precheck param

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java [hunk 4]
```diff
@@ -105,56 +96,59 @@ public int read(byte[] b) throws IOException {
...
-  private void openStream() throws IOException {
-    if (mIn != null) { // stream is already open
-      return;
-    }
-    GetObjectRequest getReq = new GetObjectRequest(mBucketName, mKey);
+  private S3Object getObject(long start, long end) {
     // If the position is 0, setting range is redundant and causes an error if the file is 0 length
-    if (mPos > 0) {
-      getReq.setRange(mPos);
+    if (start > 0 || end > 0) {
```
Comment (jja725) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:124]: if start=0 and end =1, this would error out for file with 0 length?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java [hunk 5]
```diff
@@ -105,56 +96,59 @@ public int read(byte[] b) throws IOException {
 
   @Override
   public int read(byte[] b, int offset, int length) throws IOException {
-    if (length == 0) {
-      return 0;
+    S3Object object = getObject(mPos, mPos + length);
+    if (object == null) {
+      // range request cannot meet
+      object = getObject(mPos);
```
Comment (jja725) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:102]: could you elaborate what's the case this request could not meet. I fee like this might be covered by precondition check

**Review by jja725 — COMMENTED** [review]
Thanks for the contribution! And we need microbench to see the performance benefits against naive S3ObjectInputStream.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 2 above)
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:1474]: Agree! but i am a little worried about different access patterns and data set may be good for different buffer size. But it will be good if we have a "always close to optimal" buffer size as default and no related property key

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java (hunk 1 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:89]: because single read() returns value

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java (hunk 3 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:99]: Will add unit test to understand the corner cases better

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java (hunk 4 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:124]: will add unit test to make sure it works

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java (hunk 5 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:102]: need to do more testing to understand, that's from their getObject() API description. If not meet request, will return null. They may return null for different reasons (e.g. wrong bucket name, wrong key)

**Comment by LuQQiu** [comment]
Next steps:
1. Add seek() to S3AInputStream
2. Add buffered seekable input stream similar to https://github.com/facebookarchive/hadoop-20/blob/master/src/core/org/apache/hadoop/fs/BufferedFSInputStream.java
Use for S3 for testing now, but will consider moving to general client
4. General seekable input stream test
5. General buffered seekable input stream test
6. FIO end to end performance benchmarking

**Comment by LuQQiu** [comment]
FUSE FIO with S3 range read performance
Original master - 23.4MiB/s (24.5MB/s)
New Fuse FIO with s3 range read
<google-sheets-html-origin><style type="text/css"></style>

Buffer size | Alluxio-Fuse S3 Performance |  
-- | -- | --
1MB | 34.0MiB/s (36.7MB/s)33.1MiB/s (34.7MB/s) |  
2MB | 54.2MiB/s (56.8MB/s)65.6MiB/s (68.8MB/s) |  
4MB | 70.5MiB/s (73.9MB/s)80.2MiB/s (84.1MB/s) |  
40MB | 66.2MiB/s (69.4MB/s)87.9MiB/s (92.1MB/s) |  
100MB | 81.1MiB/s (85.1MB/s)90.1MiB/s (94.5MB/s) |  
150MB | 91.1MiB/s (95.5MB/s)90.6MiB/s (95.0MB/s) |  

Test tool: FIO sequential read 1GB file
fio --name=sequentialread --rw=read --bs=4m --numjobs=1 --filesize=1g -direct=1 --group_reporting --nrfiles 1

**Comment by LuQQiu** [comment]
For dora branch the changes are ready.

Some other works if need to merge to master
- 1. where to add the input stream buffer to in old path
- 2. General seekable input stream test
General buffered seekable input stream test

**Comment by LuQQiu** [comment]
@jja725 PTAL, thanks!

File: core/common/src/main/java/alluxio/file/SeekableBufferedInputStream.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+public class SeekableBufferedInputStream extends BufferedInputStream implements Seekable {
+
+  /**
+   * Constructor.
+   * @param in input stream implements seekable
+   * @param size the buffer size
+   */
+  public SeekableBufferedInputStream(InputStream in, int size) {
+    super(in, size);
+    Preconditions.checkArgument(in instanceof Seekable,
```
Comment (jja725) [inline core/common/src/main/java/alluxio/file/SeekableBufferedInputStream.java:34]: nit: check before super

File: integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInStream.java [hunk 7]
```diff
@@ -110,19 +110,19 @@ public synchronized int read(ByteBuffer buf, long size, long offset) {
     }
     final int sz = (int) size;
     int totalRead = 0;
-    int currentRead = 0;
+    int currentRead;
     try {
       mInStream.seek(offset);
-      while (currentRead >= 0 && totalRead < sz) {
+      do {
```
Comment (jja725) [inline integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInStream.java:116]: just wonder why we change some while loop to do while?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java [hunk 8]
```diff
@@ -105,56 +86,79 @@ public int read(byte[] b) throws IOException {
...
+        if (!lastException.isRetryable()) {
+          throw lastException;
+        }
+      }
+    } while (mRetryPolicy.attempt());
+    if (object == null) {
+      if (lastException != null) {
+        throw lastException;
+      }
+      // should not reach here
```
Comment (jja725) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java:126]: delete if never reach here

**Review by jja725 — APPROVED** [review]
leave some minor comments but feel free to merge it when you feel comfortable

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java [hunk 9]
```diff
@@ -162,6 +170,12 @@ private void updateStreamIfNeeded() {
     if (mUfsInStream.isPresent()) {
       return;
     }
-    mUfsInStream = Optional.of(mFileOpener.apply(mPosition));
+    InputStream ufsInStream = mFileOpener.apply(mPosition);
```
Comment (jja725) [inline core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java:173]: I'm still a bit confused why we need to pass fileOpener instead of just have a function inside?

File: core/common/src/main/java/alluxio/file/SeekableBufferedInputStream.java (hunk 6 above)
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/file/SeekableBufferedInputStream.java:34]: in constructor, they require that super() in the first line before all other logics (including preconditions.check)

I was also confusing why this is enforced that way:
https://stackoverflow.com/questions/1168345/why-do-this-and-super-have-to-be-the-first-statement-in-a-constructor

File: integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInStream.java (hunk 7 above)
Comment (LuQQiu) [reply integration/fuse/src/main/java/alluxio/fuse/file/FuseFileInStream.java:116]: int currentRead = 0 
- while(currentRead >=0) currentRead = instream.read()
- do currentRead = instream.read while (currentRead > 0)
The first approach is correct in normal cases, just afraid that some instream.read() returns 0 forever and causes we stuck in the while loop forever (although normally it wont)
The second approach can make sure we do not continue read while inStream.read returns 0 or -1

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java (hunk 9 above)
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java:173]: File Opener is a function `Function<Long, InputStream>` to open the file.
This is for convenience, or we can pass in `UnderFileSystem` and file path and add a `openFile(offset)` function inside the UfsFileInStream.
Both approaches work, i do not see which one is much better than another
```
      return new UfsFileInStream(offset -> {
        try {
          return mUfs.get().open(status.getPath(), OpenOptions.defaults().setOffset(offset));
        } catch (IOException e) {
          throw AlluxioRuntimeException.from(e);
        }
      }, status.getLength());
    });
```

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java [hunk 10]
```diff
@@ -79,17 +85,11 @@ public int read(byte[] b, int off, int len) throws IOException {
...
-    int totalRead = 0;
-    while (totalRead < len) {
-      currentRead = mUfsInStream.get().read(b, off + totalRead, len - totalRead);
-      if (currentRead <= 0) {
-        break;
-      }
-      totalRead += currentRead;
+    int bytesRead = mUfsInStream.get().read(b, off, len);
+    if (bytesRead > 0) {
+      mPosition += bytesRead;
```
Comment (huanghua78) [inline core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java:90]: This "try-to-read-as-much-as-possible" while-block is moved to underfs/s3a, right?
What if we are using some other underfs?
